### PR TITLE
ci: add local changelog creation back

### PR DESCRIPTION
**Pull request by randi274**
### What does this PR do?
Creates a task to update the changelog for the current release.

### What issues does this PR fix or reference?
@W-13097185@

### Functionality Before
Changelog creation was a manual process, having been broken in GHA for a while. Example of a successful run [here](https://github.com/forcedotcom/salesforcedx-vscode/compare/develop...release/v57.15.0).

### Functionality After
To create a changelog:
- Open up VS Code Insiders (or standard VS Code)
- Open the command palette
- Tasks:Run Task > `Create Change Log`
- (Optionally) Specify a release version. If left empty, the task will grab the latest release branch with the `release/v` prefix.
- If there are changes, our CHANGELOG.md will be updated and opened to be edit-ready.
- If there are no changes worth documenting, the task will print out a message that indicates we can skip the release. 

Tips:
- run a `git pull` first to grab the latest version of the release branch locally
- if using Insiders, make sure you have `code-insiders` installed in your PATH to have the file open there
Screenshot 2023-05-22 at 7 39 21 PM

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -57,6 +57,7 @@
     "bootstrap": "lerna bootstrap --no-ci -- --no-package-lock",
     "commit-init": "commitizen init cz-conventional-changelog --save-dev --save-exact --force",
     "commit": "git-cz",
+    "changelog": "npx ts-node ./scripts/create-release-notes.ts",
```
Comment (randi274) [inline package.json:60]: I wasn't able to get `npx` to cooperate in the `tasks.json` so I added a script here instead.

File: scripts/change-log-generator-utils.js [hunk 2]
```diff
@@ -283,6 +285,32 @@ function getReleaseDate() {
     }).format(releaseDate);
 }
 
+/**
+ *
+ * Complete the heavy lifting to update the changelog by grabbing the
+ * new commits, grouping everything, and creating the text for editing.
+ * @param {string} remoteReleaseBranch
+ * @param {string} remotePreviousBranch
```
Comment (randi274) [inline scripts/change-log-generator-utils.js:293]: Comparing both of these on remote was more consistent for if I didn't have the branch locally for instance.

File: scripts/create-release-notes.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  if (!(releaseBranch && constants.RELEASE_REGEX.exec(releaseBranch))) {
+    console.log(
+      "Invalid release '" + releaseBranch + "'. Expected format [xx.yy.z]."
+    );
+    process.exit(-1);
+  }
+}
+
+const releaseBranchName = getRemoteReleaseBranch();
+logger(`\nStep 2: Getting latest tag to compare last published version`);
```
Comment (randi274) [inline scripts/create-release-notes.ts:63]: There's probably a smarter way to keep track of the steps, but I promise they're all in order :)

File: contributing/publishing.md [hunk 4]
```diff
@@ -176,16 +178,19 @@ from Atlassian on the flow. These steps are manual because you might encounter m
 1. `git merge release/vxx.y.z`
 1. `git push`
 
-## Manual Publish in Open VSX Registry
```
Comment (randi274) [inline contributing/publishing.md:179]: Automatic spacing updates follow below

File: scripts/change-log-generator-utils.js [hunk 5]
```diff
@@ -283,6 +285,32 @@ function getReleaseDate() {
...
+    writeChangeLog(changeLog);
+  } else {
+    console.log(`No commits found, so we can skip this week's release. Carry on!`);
+    process.exit(0);
+  }
+}
+
 module.exports = {
-  getPreviousReleaseBranch, parseCommits, getMessagesGroupedByPackage, getChangeLogText, getCommits, writeChangeLog
+  updateChangeLog
```
Comment (randi274) [inline scripts/change-log-generator-utils.js:315]: These were previously all being referenced in `create-release-branch.js` and I opted to simplify it all to a single function that needed exporting.

File: package.json (hunk 1 above)
Comment (gbockus-sf) [reply package.json:60]: You shouldn't need npx here.  npx is how you get access to the npm state from the command line. B/c we're running in an npm script it'll know where to find ts-node. 
```   
 "changelog": "ts-node ./scripts/create-release-notes.ts",
```

File: scripts/change-log-generator-utils.js [hunk 6]
```diff
@@ -283,6 +285,32 @@ function getReleaseDate() {
     }).format(releaseDate);
 }
 
+/**
+ *
+ * Complete the heavy lifting to update the changelog by grabbing the
+ * new commits, grouping everything, and creating the text for editing.
+ * @param {string} remoteReleaseBranch
+ * @param {string} remotePreviousBranch
+ * @returns
```
Comment (gbockus-sf) [inline scripts/change-log-generator-utils.js:294]: no returns annotation necessary.

File: scripts/create-release-notes.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
+import * as shell from 'shelljs';
+const changeLogGeneratorUtils = require('./change-log-generator-utils');
+const constants = require('./change-log-constants');
+
+const [_, __, releaseOverride] = process.argv
```
[the remark this answers is not in the document]
Comment (gbockus-sf) [reply scripts/create-release-notes.ts:5]: does prettier need to get run? There's no semi on this line

File: scripts/create-release-notes.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
+import * as shell from 'shelljs';
+const changeLogGeneratorUtils = require('./change-log-generator-utils');
+const constants = require('./change-log-constants');
```
Comment (gbockus-sf) [inline scripts/create-release-notes.ts:3]: :high_five: writing this in typescript.  You can use imports here

File: scripts/create-release-notes.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * Checks if the user has provided a release branch override. If they
+ * have not, return the latest release branch.
+ */
+function getCurrentRemoteReleaseBranch(): string {
+  logger('\nStep 1: Determine release branch.');
+  let releaseBranch;
+  if (!releaseOverride) {
+    console.log(`releaseOverride not provided. Getting latest release.`);
```
Comment (gbockus-sf) [inline scripts/create-release-notes.ts:23]: you made an awesome logger then used console.log :P

File: scripts/create-release-notes.ts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * Confirms the release branch is in the correct format
+ */
+function validateReleaseBranch(releaseBranch): void {
+  if (!(releaseBranch && constants.RELEASE_REGEX.exec(releaseBranch))) {
+    console.log(
+      "Invalid release '" + releaseBranch + "'. Expected format [xx.yy.z]."
+    );
+    process.exit(-1);
```
Comment (gbockus-sf) [inline scripts/create-release-notes.ts:68]: nit: -1 is a weird exit code. Should this just be 1?

**Review by gbockus-sf — APPROVED** [review]
Couple of simple suggestions in there, Def leaving it better then you found it.

File: package.json (hunk 1 above)
Comment (randi274) [reply package.json:60]: ah, awesome!

File: scripts/create-release-notes.ts (hunk 10 above)
Comment (randi274) [reply scripts/create-release-notes.ts:68]: Probably - this was all over the validation-utils.js and create-release-branch.js so just a copypasta situation

File: scripts/create-release-notes.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (randi274) [reply scripts/create-release-notes.ts:5]: for the first - a fun stackoverflow find to get right at releaseOverride. the first _ ends up being the path to ts-node, and the second __ the path to the specific script (create-release-notes)
neither var
only single underscore

File: scripts/create-release-notes.ts (hunk 9 above)
Comment (randi274) [reply scripts/create-release-notes.ts:23]: hahahaha I know I know - but I decided I didn't always love the extra *** at the start of the line, and liked it more to separate steps. I do need to add a logger to change-log-generator-utils for that at least to be consistent though.

**Pull request merged**